File: leetcode/1-100/70-climbing-stairs/bench/climbing_stairs.c
/*
 * Benchmark workload — Climbing Stairs (LeetCode #70).
 * C mirror of bench/climbing_stairs.kara. The ★'s two-counter Fibonacci
 * recurrence run K=30_000_000 times over a sweep of n = 1 + k%45, folding each
 * result into a rolling polynomial hash. No allocation — a pure integer-add /
 * branch benchmark of the recurrence loop. See ../README.md § Benchmarks.
 */
#include <stdio.h>
#include <stdint.h>

static int64_t climb(int64_t n) {
    if (n <= 2) return n;
    int64_t a = 1, b = 2;
    for (int64_t i = 3; i <= n; i++) {
        int64_t next = a + b;
        a = b;
        b = next;
    }
    return b;
}

int main(void) {
    const int64_t total = 30000000;
    const int64_t modulus = 1000000007;
    const int64_t span = 45;

    int64_t acc = 0;
    for (int64_t k = 0; k < total; k++) {
        int64_t n = 1 + (k % span);
        acc = (acc * 131 + climb(n)) % modulus;
    }
    printf("%lld\n", (long long)acc);
    return 0;
}

File: leetcode/1-100/70-climbing-stairs/bench/climbing_stairs.rs
// Benchmark workload — Climbing Stairs (LeetCode #70).
// Rust mirror of bench/climbing_stairs.kara. The ★'s two-counter Fibonacci
// recurrence run K=30_000_000 times over a sweep of n = 1 + k%45, folding each
// result into a rolling polynomial hash. No allocation — a pure integer-add /
// branch benchmark of the recurrence loop. Compiled with `rustc -O`.
// See ../README.md § Benchmarks.

fn climb(n: i64) -> i64 {
    if n <= 2 {
        return n;
    }
    let mut a = 1i64;
    let mut b = 2i64;
    let mut i = 3i64;
    while i <= n {
        let next = a + b;
        a = b;
        b = next;
        i += 1;
    }
    b
}

fn main() {
    let total: i64 = 30_000_000;
    let modulus: i64 = 1_000_000_007;
    let span: i64 = 45;

    let mut acc: i64 = 0;
    for k in 0..total {
        let n = 1 + (k % span);
        acc = (acc * 131 + std::hint::black_box(climb(n))) % modulus;
    }
    println!("{}", acc);
}
